L1Trigger/TrackerDTC/src/Setup.cc

```diff
@@ -0,0 +1,730 @@
+    windowSizeTiltedLayerRings_.reserve(pSetsTiltedLayer.size());
+    for (const auto& pSet : pSetsTiltedLayer)
+      windowSizeTiltedLayerRings_.emplace_back(pSet.getParameter<vector<double>>("TiltedCut"));
+    windowSizeEndcapDisksRings_.reserve(pSetsEncapDisks.size());
+    for (const auto& pSet : pSetsEncapDisks)
+      windowSizeEndcapDisksRings_.emplace_back(pSet.getParameter<vector<double>>("EndcapCut"));
+    maxWindowSize_ = -1;
+    for (const auto& windowss : {windowSizeTiltedLayerRings_, windowSizeEndcapDisksRings_, {windowSizeBarrelLayers_}})
+      for (const auto& windows : windowss)
+        for (const auto& window : windows)
+          maxWindowSize_ = max(maxWindowSize_, (int)(window / baseWindowSize_));
+  }
+
+  // create bend encodings
+  void Setup::encodeBend(vector<vector<double>>& encodings, bool ps) const {
+    for (int window = 0; window < maxWindowSize_ + 1; window++) {
+      set<double> encoding;
+      for (int bend = 0; bend < window + 1; bend++)
+        encoding.insert(stubAlgorithm_->degradeBend(ps, window, bend));
+      encodings.emplace_back(encoding.begin(), encoding.end());
+    }
+  }
+
+  // create encodingsLayerId
+  void Setup::encodeLayerId() {
+    vector<vector<DTCELinkId>> dtcELinkIds(numDTCs_);
+    for (vector<DTCELinkId>& dtcELinkId : dtcELinkIds)
+      dtcELinkId.reserve(numModulesPerDTC_);
+    for (const DTCELinkId& dtcLinkId : cablingMap_->getKnownDTCELinkIds())
+      dtcELinkIds[dtcId(dtcLinkId.dtc_id())].push_back(dtcLinkId);
+    for (int dtcBoard = 0; dtcBoard < numDTCsPerRegion_; dtcBoard++) {
+      set<int> encodingLayerId;
+      for (int region = 0; region < numRegions_; region++) {
+        const int dtcId = region * numDTCsPerRegion_ + dtcBoard;
+        for (const DTCELinkId& dtcLinkId : dtcELinkIds[dtcId]) {
+          const DetId& detId = cablingMap_->dtcELinkIdToDetId(dtcLinkId)->second;
+          const bool barrel = detId.subdetId() == StripSubdetector::TOB;
+          const int layerId =
+              barrel ? trackerTopology_->layer(detId) : trackerTopology_->tidWheel(detId) + offsetLayerDisks_;
+          encodingLayerId.insert(layerId);
+        }
+      }
+      // check configuration
+      if ((int)encodingLayerId.size() > hybridNumLayers_) {
+        cms::Exception exception("overflow");
+        exception << "Cabling map connects more than " << hybridNumLayers_ << " layers to a DTC.";
+        exception.addContext("tt::Setup::Setup");
+        throw exception;
+      }
+      encodingsLayerId_.emplace_back(encodingLayerId.begin(), encodingLayerId.end());
+    }
+  }
+
+  // create sensor modules
+  void Setup::produceSensorModules() {
+    sensorModules_.reserve(numModules_);
+    dtcModules_ = vector<vector<SensorModule*>>(numDTCs_);
+    for (vector<SensorModule*>& dtcModules : dtcModules_)
+      dtcModules.reserve(numModulesPerDTC_);
+    enum SubDetId { pixelBarrel = 1, pixelDisks = 2 };
```

Comment:
why not use the existing enum at https://github.com/cms-sw/cmssw/blob/master/DataFormats/SiPixelDetId/interface/PixelSubdetector.h?